File: php/tests/routing.test.php
<?php
// Distance-vector routing tests (protocol.md §5). The algorithm and its
// constants mirror the Java, Elixir, Go, Rust, and JS routers; agreement is
// what lets a PHP node relay in a mixed mesh. (The relay path itself is a
// blocking single-loop concern proven by the interop matrix and tier 8, not an
// in-process test — PHP cannot run several nodes in one process.)
use Bonemesh\RoutingTable;
use Bonemesh\Dedup;

test('neighbor is its own next hop and is advertised', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('B', 10);
    assertEq('b', $t->nextHop('b'));
    assertEq(10, $t->advertiseTo('x')['b']);
});

test('learn install, refresh, cheaper', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('b', 10);
    $t->observeNeighbor('d', 100);
    $t->learnRoute('c', 'b', 5); // 5 + 10 = 15 via b
    assertEq('b', $t->nextHop('c'));
    $t->learnRoute('c', 'd', 1); // 1 + 100 = 101 — worse, ignored
    assertEq('b', $t->nextHop('c'));
    $t->observeNeighbor('e', 1);
    $t->learnRoute('c', 'e', 2); // 2 + 1 = 3 via e — cheaper
    assertEq('e', $t->nextHop('c'));
    $t->learnRoute('c', 'e', 500); // same via — refresh
    assertEq(501, $t->advertiseTo('x')['c']);
});

test('learn guards', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('b', 10);
    $t->learnRoute('self', 'b', 1);
    $t->learnRoute('b', 'b', 1);
    $t->learnRoute('c', 'z', 1);
    assertEq(0, count($t->routeTable()));
});

test('poison from the route\'s own next hop withdraws it', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('b', 10);
    $t->learnRoute('c', 'b', 5);
    $t->learnRoute('c', 'b', 1000000000); // Elixir/JS sentinel
    assertNull($t->nextHop('c'));
});

test('poison from another neighbor is a no-op', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('b', 10);
    $t->observeNeighbor('d', 10);
    $t->learnRoute('c', 'b', 5);
    $t->learnRoute('c', 'd', RoutingTable::UNREACHABLE);
    assertEq('b', $t->nextHop('c'));
});

test('advertise with split-horizon poisoned reverse', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('b', 10);
    $t->learnRoute('c', 'b', 5);
    $toB = $t->advertiseTo('b');
    assertTrue($toB['c'] >= RoutingTable::POISON_THRESHOLD);
    assertTrue(!isset($toB['b']));
    $toX = $t->advertiseTo('x');
    assertEq(15, $toX['c']);
    assertEq(10, $toX['b']);
});

test('remove neighbor withdraws its routes', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('b', 10);
    $t->learnRoute('c', 'b', 5);
    $t->removeNeighbor('b');
    assertNull($t->nextHop('c'));
    assertNull($t->nextHop('b'));
});

test('EWMA smoothing (alpha 0.2)', function () {
    $t = new RoutingTable('self');
    $t->observeNeighbor('b', 100);
    $t->observeNeighbor('b', 0); // 0.2*0 + 0.8*100 = 80
    assertEq(80, $t->advertiseTo('x')['b']);
});

test('dedup is bounded', function () {
    $d = new Dedup(2);
    assertTrue(!$d->sawBefore('a'));
    assertTrue($d->sawBefore('a'));
    $d->sawBefore('b');
    $d->sawBefore('c'); // evicts a
    assertTrue(!$d->sawBefore('a'));
});

File: rust/tests/routing_test.rs
// Distance-vector routing tests (protocol.md §5). The algorithm and its wire
// constants mirror the Java and Elixir reference routers; agreement is what lets
// a Rust node relay in a mixed mesh.
use bonemesh::routing::{Dedup, Table, POISON_THRESHOLD, UNREACHABLE};

#[test]
fn neighbor_is_its_own_next_hop_and_advertised() {
    let mut t = Table::new("self");
    t.observe_neighbor("B", 10);
    assert_eq!(t.next_hop("b").as_deref(), Some("b"));
    assert_eq!(t.advertise_to("x")["b"].as_i64(), Some(10));
}

#[test]
fn learn_install_refresh_cheaper() {
    let mut t = Table::new("self");
    t.observe_neighbor("b", 10);
    t.observe_neighbor("d", 100);
    t.learn_route("c", "b", 5); // 5 + 10 = 15 via b
    assert_eq!(t.next_hop("c").as_deref(), Some("b"));
    t.learn_route("c", "d", 1); // 1 + 100 = 101 via d — worse, ignored
    assert_eq!(t.next_hop("c").as_deref(), Some("b"));
    t.observe_neighbor("e", 1);
    t.learn_route("c", "e", 2); // 2 + 1 = 3 via e — cheaper, replaces
    assert_eq!(t.next_hop("c").as_deref(), Some("e"));
    t.learn_route("c", "e", 500); // same via — refresh cost to 501
    assert_eq!(t.advertise_to("x")["c"].as_i64(), Some(501));
}

#[test]
fn learn_guards() {
    let mut t = Table::new("self");
    t.observe_neighbor("b", 10);
    t.learn_route("self", "b", 1);
    t.learn_route("b", "b", 1);
    t.learn_route("c", "z", 1); // via not a neighbor
    assert!(t.route_table().is_empty());
}

#[test]
fn poison_from_own_next_hop_withdraws() {
    let mut t = Table::new("self");
    t.observe_neighbor("b", 10);
    t.learn_route("c", "b", 5);
    t.learn_route("c", "b", 1_000_000_000); // Elixir's sentinel counts as poison
    assert_eq!(t.next_hop("c"), None);
}

#[test]
fn poison_from_other_neighbor_is_noop() {
    let mut t = Table::new("self");
    t.observe_neighbor("b", 10);
    t.observe_neighbor("d", 10);
    t.learn_route("c", "b", 5);
    t.learn_route("c", "d", UNREACHABLE);
    assert_eq!(t.next_hop("c").as_deref(), Some("b"));
}

#[test]
fn advertise_poisoned_reverse() {
    let mut t = Table::new("self");
    t.observe_neighbor("b", 10);
    t.learn_route("c", "b", 5);
    let to_b = t.advertise_to("b");
    assert!(to_b["c"].as_i64().unwrap() >= POISON_THRESHOLD);
    assert!(to_b.get("b").is_none());
    let to_x = t.advertise_to("x");
    assert_eq!(to_x["c"].as_i64(), Some(15));
    assert_eq!(to_x["b"].as_i64(), Some(10));
}

#[test]
fn remove_neighbor_withdraws_its_routes() {
    let mut t = Table::new("self");
    t.observe_neighbor("b", 10);
    t.learn_route("c", "b", 5);
    t.remove_neighbor("b");
    assert_eq!(t.next_hop("c"), None);
    assert_eq!(t.next_hop("b"), None);
}

#[test]
fn ewma_smoothing() {
    let mut t = Table::new("self");
    t.observe_neighbor("b", 100);
    t.observe_neighbor("b", 0); // 0.2*0 + 0.8*100 = 80
    assert_eq!(t.advertise_to("x")["b"].as_i64(), Some(80));
}

#[test]
fn dedup_bounded() {
    let mut d = Dedup::new(2);
    assert!(!d.seen("a"));
    assert!(d.seen("a"));
    d.seen("b");
    d.seen("c"); // evicts a
    assert!(!d.seen("a"));
}
